### DISCUSSION TITLE
ValidationError: Entity of type A expected for property B.as, B {} of type object given.

### DISCUSSION THREAD

**Discussion by RoyLee** [mikro-orm/mikro-orm#2573] [Q&A]
**Describe the bug**
I am receiving a `ValidationError` error when trying to add a related object, `book`, to my `author` entity object. Can someone please help me understand why this is happening?

Code I am running:
```
const author = await em.findOneOrFail(Author, { name: "George" })
const book = new Book("test");
author.books.add(book); // errors on this line
```

Note:
I've defined my entity relationships like so:
`@ManyToOne('Author')`
`@OneToMany('Book', 'author')`
Because of the following error: [ReferenceError: Cannot access 'Author' before initialization
](https://github.com/mikro-orm/mikro-orm/issues/1310)

**Stack trace**
```
ValidationError: Entity of type Book expected for property Author.books, Author {
  _id: undefined,
  id: undefined,
  createdAt: ISODate('2021-12-24T22:38:35.896Z'),
  updatedAt: ISODate('2021-12-24T22:38:35.896Z'),
  body: 'test',
  author: undefined
} of type object given. If you are using Object.assign(entity, data), use em.assign(entity, data) instead.
    at Function.notEntity (/Users/me/projects/my-repo/node_modules/@mikro-orm/core/errors.js:40:16)
    at Collection.validateItemType (/Users/me/projects/my-repo/node_modules/@mikro-orm/core/entity/Collection.js:317:44)
    at /Users/me/projects/my-repo/node_modules/@mikro-orm/core/entity/Collection.js:111:40
    at Array.forEach (<anonymous>)
    at Collection.add (/Users/me/projects/my-repo/node_modules/@mikro-orm/core/entity/Collection.js:111:19)
    at __WEBPACK_DEFAULT_EXPORT__ (webpack-internal:///./src/pages/api/guild/create.ts:30:34)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at Object.apiResolver (/Users/me/projects/my-repo/node_modules/next/server/api-utils.ts:107:5)
    at DevServer.handleApiRequest (/Users/me/projects/my-repo/node_modules/next/server/next-server.ts:1515:5) {
  entity: undefined
}
```

**To Reproduce**

Book.ts
```
@Entity({ tableName: "books" })
export class Book {
  @Property()
  body?: string;

  @ManyToOne('Author')
  author!: Author;

  constructor(body: string) {
    super();
    this.body = body;
  }
}
```

Author.ts
```
@Entity({ tableName: "authors" })
export class Author {
  @Property()
  @Unique()
  name: string;

  @OneToMany('Book', 'author')
  books = new Collection<Book>(this);

  constructor(name: string) {
    super();
    this.name = name;
  }
}
```

config.ts
```
const config: Options = {
  type: "mongo",
  clientUrl: process.env.MONGODB_URI,
  entities: [
    Author,
    Book,
  ],
  debug: process.env.NODE_ENV === "development",
  highlighter: new MongoHighlighter(),
  metadataProvider: ReflectMetadataProvider,
  ensureIndexes: true,
  pool: { min: 10, max: 100 },
};
export default config;
```

**Expected behavior**
Expect the book to be added to the author relationship collection.

**Versions**
| Dependency | Version |
| node | 17.0.1 |
| typescript | 4.3.5 |
| mikro-orm | 4.5.9 |
| mongodb / your-driver | 4.0.0 |

**Comment by B4nan** [comment]
Can you provide full reproduction? Also that lowecase book class in your example is a typo? Should be `new Book()`, not `new book()`. 

I am not fully sure if node 17 is supported with v4, might as well be something with ts config (e.g. v4 does not work with esnext). But sounds weird, examples like that are part of the test suite. Try logging the entity before you add it to the collection. 

I could also imagine something like that if you import from bad location, that error suggests it is not a discovered entity. That might happen if you import from dist I guess.

**Comment by RoyLee** [comment]
Hi B4nan,
Thanks for responding so quick and for catching the typo. I've fixed it by making it uppercase: `new Book()`. 

Is there documentation on what node version I should use? I didn't see it in the repo. I can downgrade if necessary.

You are correct that I am importing from dist though according to the typescript config:
tsconfig.server.json:
```
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "module": "commonjs",
    "outDir": "dist",
    "target": "es2017",
    "isolatedModules": false,
    "noEmit": false,
    "declaration": true,
    "emitDecoratorMetadata": true
  },
  "include": [
    "src/server/**/*.ts"
  ]
}
```
What do you need for the full reproduction? I've given most of the code. What else would you like to see?

**Comment by B4nan** [comment]
Full (but minimal) project i can run and see the failure myself (or even better a test case). 

I dont think you should be importing from dist, so i would expect that to be your problem.

**Comment by RoyLee** [comment]
Sounds good @B4nan . I'll work on creating a minimal project that you can look at. Will message it here when I have it ready for you to look at. Thanks!
